aspects/aspects.go

```diff
@@ -0,0 +1,343 @@
+
+// Schema takes in data from the DataBag and validates that it's valid and could
+// be committed.
+type Schema interface {
+	Validate(data []byte) error
+}
+
+// Directory holds a series of related aspects.
+type Directory struct {
+	Name    string
+	dataBag DataBag
+	schema  Schema
+	aspects map[string]*Aspect
+}
+
+// NewAspectDirectory returns a new aspect directory for the following aspects
+// and access patterns.
+func NewAspectDirectory(name string, aspects map[string]interface{}, dataBag DataBag, schema Schema) (*Directory, error) {
+	if len(aspects) == 0 {
+		return nil, errors.New(`cannot create aspects directory: no aspects`)
+	}
+
+	aspectDir := Directory{
+		Name:    name,
+		dataBag: dataBag,
+		schema:  schema,
+		aspects: make(map[string]*Aspect, len(aspects)),
+	}
+
+	for name, v := range aspects {
+		aspectPatterns, ok := v.([]map[string]string)
+		if !ok {
+			return nil, errors.New("cannot create aspect: access patterns should be a list of maps")
+		} else if len(aspectPatterns) == 0 {
+			return nil, errors.New("cannot create aspect without access patterns")
+		}
+
+		aspect := &Aspect{
+			Name:           name,
+			accessPatterns: make([]*accessPattern, 0, len(aspectPatterns)),
+			directory:      aspectDir,
+		}
+
+		for _, aspectPattern := range aspectPatterns {
+			name, ok := aspectPattern["name"]
+			if !ok || name == "" {
+				return nil, errors.New(`cannot create aspect pattern without a "name" field`)
+			}
+
+			// TODO: validate that a path isn't a subset of another (possibly somewhere else).
+			// Otherwise,  we can write a user value in a subkey of a path (that should be a map).
+			path, ok := aspectPattern["path"]
+			if !ok || path == "" {
+				return nil, errors.New(`cannot create aspect pattern without a "path" field`)
+			}
+
+			access, err := newAccessType(aspectPattern["access"])
+			if err != nil {
+				return nil, fmt.Errorf("cannot create aspect pattern: %w", err)
+			}
+
+			aspect.accessPatterns = append(aspect.accessPatterns, &accessPattern{
+				name:   name,
+				path:   path,
+				access: access,
+			})
+		}
+
+		aspectDir.aspects[name] = aspect
+	}
+
+	return &aspectDir, nil
+}
+
+// Aspect returns an aspect from the aspect directory.
+func (d *Directory) Aspect(aspect string) *Aspect {
+	return d.aspects[aspect]
+}
+
+// Aspect is a group of access patterns under a directory.
+type Aspect struct {
+	Name           string
+	accessPatterns []*accessPattern
+	directory      Directory
+}
+
+// Set sets the named aspect to a specified value.
+func (a *Aspect) Set(name string, value interface{}) error {
+	for _, p := range a.accessPatterns {
+		if p.name != name {
+			continue
+		}
+
+		if !p.isWriteable() {
+			return fmt.Errorf("cannot set %q: path is not writeable", name)
+		}
+
+		if err := a.directory.dataBag.Set(p.path, value); err != nil {
+			return err
+		}
+
+		data, err := a.directory.dataBag.Data()
+		if err != nil {
+			return err
+		}
+
+		return a.directory.schema.Validate(data)
```

Comment:
Silly question - should we either vaidate before setting the data in the data bag or revert in case the validation fails?

Reply:
for now the data bag is fully in memory and detached from state, in principle you could turn data in the state into data bags and apply ops and not write back if it fails, so doing the check this way is not wrong as things stand. In practice we will implement some notion of transaction when we get that far, and adapt this if needed